src/backend/distributed/transaction/distributed_deadlock_detection.c

```diff
@@ -25,11 +25,264 @@
+											   keysize);
+static int DistributedTransactionIdCompare(const void *a, const void *b);
+
+
+PG_FUNCTION_INFO_V1(check_distributed_deadlocks);
+
+
+/*
+ * check_distributed_deadlocks is the external API for manually
+ * checking for distributed deadlocks. For the details, see
+ * CheckForDistributedDeadlocks().
+ */
+Datum
+check_distributed_deadlocks(PG_FUNCTION_ARGS)
+{
+	bool deadlockFound = CheckForDistributedDeadlocks();
+
+	return BoolGetDatum(deadlockFound);
+}
+
+
+/*
+ * CheckForDistributedDeadlocks is the entry point for detecing
+ * distributed deadlocks.
+ *
+ * In plain words, the function first builds a wait graph by
+ * adding the wait edges from the local node and then adding the
+ * remote wait edges to form a global wait graph. Later, the wait
+ * graph is converted into another graph representation (adjacency
+ * lists) for more efficient searches. Finally, a DFS is done on
+ * the adjacency lists. Finding a cycle in the graph unveils a
+ * distributed deadlock. Upon finding a deadlock, the youngest
+ * participant backend is killed.
+ *
+ * The function returns true if a deadlock is found. Otherwise, returns
+ * false.
+ */
+bool
+CheckForDistributedDeadlocks(void)
+{
+	WaitGraph *waitGraph = BuildGlobalWaitGraph();
+	HTAB *adjacencyList = BuildAdjacencyListsForWaitGraph(waitGraph);
+	HASH_SEQ_STATUS status;
+	TransactionNode *transactionNode = NULL;
+
+	/*
+	 * We iterate on transaction nodes and search for deadlocks where the
+	 * starting node is the given transaction node.
+	 */
+	hash_seq_init(&status, adjacencyList);
+	while ((transactionNode = (TransactionNode *) hash_seq_search(&status)) != 0)
+	{
+		bool deadlockFound = false;
+		List *deadlockPath = NIL;
+
+		ResetVisitedFields(adjacencyList);
+
+		deadlockFound = CheckDeadlockForTransactionNode(transactionNode, &deadlockPath);
+		if (deadlockFound)
+		{
+			TransactionNode *youngestTransaction = transactionNode;
+			ListCell *participantTransactionCell = NULL;
+
+			/* there should be at least two transactions to get into a deadlock */
+			Assert(list_length(deadlockPath) > 1);
+
+			/*
+			 * We search for the youngest participant for two reasons
+			 * (i) predictable results (ii) kill the youngest transaction
+			 * (i.e., if a DDL continues for 1 hour and deadlocks with a
+			 * SELECT continues for 10 msec, we prefer to kill the SELECT).
+			 */
+			foreach(participantTransactionCell, deadlockPath)
+			{
+				TransactionNode *currentNode =
+					(TransactionNode *) lfirst(participantTransactionCell);
+
+				AssocateDistributedTransactionWithBackendProc(currentNode);
+
+				if (youngestTransaction->transactionId.timestamp >
+					currentNode->transactionId.timestamp)
+				{
+					youngestTransaction = currentNode;
+				}
+			}
+
+			/* we should find the backend */
+			Assert(youngestTransaction->initiatorProc != NULL);
+
+			KillBackendDueToDeadlock(youngestTransaction->initiatorProc);
+
+			hash_seq_term(&status);
+
+			return true;
```

Comment:
I don't think we should return here. We really should search for multiple deadlocks in one go, otherwise we might never get ones that happen to be later in the hash space, if there's continually new ones "early" in the hash space.

Now that's not entirely trivial because we'll often find a deadlock coming from two different starting nodes. I think that means we need to keep a bit more state in the hash table elements. We could: a) just not reset the visited field when we found a deadlock, and move the reset to the end of `CheckDeadlockForTransactionNode`. That would avoid finding the same deadlock from different ends. b) keep a separate 'inDeadlock' field that's set when find a deadlock that then triggers a new `CheckDeadlockForTransactionNode` to fail when it re-finds such a node. c) more complex: remove cycles from the graph.